frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -588,9 +604,29 @@ pub mod pallet {
         /// However, caller have to have deposit amount.
         /// TODO: weight
         #[pallet::weight(T::WeightInfo::payout_stakers_alive_staked(T::MaxStakings::get()))]
-        pub fn register(origin: OriginFor<T>, contract_id: T::AccountId) -> DispatchResult {
-            // TODO: impls
-            Ok(())
+        pub fn register(
+            origin: OriginFor<T>,
+            contract_id: SmartContract<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            let operator = ensure_signed(origin)?;
+            let mut ledger = Self::ledger(&operator).ok_or(Error::<T>::NotController)?;
+            ensure!(
+                !RegisteredDapps::<T>::contains_key(&operator),
+                Error::<T>::AlreadyRegisteredContract
+            );
+            ensure!(
+                Self::is_contract(&contract_id),
+                Error::<T>::AddressIsNotContract
+            );
+            ensure!(
+                ledger.total >= T::RegisterDeposit::get(),
+                Error::<T>::MissingDeposit
+            );
+
+            RegisteredDapps::<T>::insert(&operator, contract_id.clone());
+            Self::deposit_event(Event::<T>::NewContract(contract_id, ledger.total));
+
+            Ok(().into())
```

Comment:
Based on the current implementation, user can 'deposit' the same amount multiple times.

E.g. if RegisterDeposit value is 100 and I have bonded 150 funds, I can register as many contracts as I want.

Reply:
please see comment on line 613-616